### PR TITLE
feat(ui): Upgrade User Profile page to UI v3

### LINKED ISSUE #411: Upgrade the User Profile page the ROCC Specification v3
The goal of this Story is to upgrade the frontend and backend of the User Profile page to meet the ROCC Specification v3.

## ROCC Specifications

- v3
  - [UI v3 Specification](https://www.figma.com/file/yAeIrg6Whc8Bqj2EnG4MWP/Challenge-Registry)
- v2
  - Legacy code developed before we moved the project to this monorepo.

## Tasks

- [ ] #316
- [x] #320
- [x] #434 
- [x] #528
- [x] #529
- [x] #530
- [x] #531 
- [x] #532
- [x] #533
- [x] #536 
- [ ] #538
- [x] #539
- [x] #695

### PR DESCRIPTION
**Pull request by vpchung**
Accomplishes #409

Edit by @tschaffter :

- Fixes #411

### CODE REVIEW

**Comment by tschaffter** [comment]
@vpchung We started documenting the figma-to-code process for the User Profile page in `docs/figma-to-code.md`.

**Comment by tschaffter** [comment]
The current issue is:

```console
Error: /__w/challenge-registry/challenge-registry/libs/challenge-registry/user-profile/src/lib/user-profile.component.scss exceeded maximum budget. Budget 8.00 kB was not met by 424 bytes with a total of 8.41 kB.
 >  NX   Running target "build" failed
   Failed tasks:
   - challenge-registry:build:production
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
Error: Process completed with exit code 1.
```

**Comment by tschaffter** [comment]
@rrchai @vpchung I fixed the issue reported by the CI workflow (see above). I merged `main` into this PR branch, which brings Nx v14 and Angular v14. This PR branch is working fine.

@rrchai Before pushing the changes we made on Friday, make sure to pull the latest commit from this PR and update your dependencies with `challenge-install`.

Here is what this branch looks like for me (does not include the changes that Rong and I did on Friday).

`nx serve challenge-registry`

Screen Shot 2022-07-24 at 09 14 12

**Comment by tschaffter** [comment]
Just updated this branch to bring new Husky hooks.

**Comment by rrchai** [comment]
Update: 
I pushed the changes made during the workshop last Friday and new changes to fix stylings. The PR is still incomplete and the styles of each sub-component is still needed to be updated. I will work on it over the week, as well as documenting important steps and issues in the `docs/figma-to-code.md`.

**Comment by tschaffter** [comment]
> I will work on it over the week, as well as documenting important steps and issues in the docs/figma-to-code.md.

Sounds like a plan.

@vpchung Does it make sense to start documenting the figma-to-teleportHQ protocol now (#197)? You could add this documentation to `docs/figma-to-code.md`.

**Comment by vpchung** [comment]
@tschaffter @rrchai yes, I am on it 👍🏼

**Comment by rrchai** [comment]
okay, I have updated the most of parts. There are a bit more adjustments than I expected. If we need to further improve on the styles, I would need more time to polish the styles and organize the docs (`figma-to-code.md`) a bit after the sage week.  Also, I haven't got time to touch on the "starred" tab yet. Should be less changes required (button color, link mocked data, etc.) since the challenge/organization card has been implemented.

 I marked it as "Ready for review" for now and @tschaffter @vpchung feel free to checkout the branch to see the current looking.

Here are two things I need help from you:
- [ ] The icons have not been implemented. Replace icon images by using the material icons, otherwise upload the icon images.
- [ ] Fix missing border/box-shadow of challenge/organization card, i.e the border color/box-shadow of `.challenge-card-container1` (I don't know how to get this information from figma)

**Comment by tschaffter** [comment]
This is fantastic work! Let's target to complete this page by our next meeting. It's exciting to now see the materialization of the Figma mockup! Thanks @rrchai and @vpchung for this amazing contribution!

I hope you have a good time off this coming week.

image

**Comment by tschaffter** [comment]
> The icons have not been implemented. Replace icon images by using the material icons, otherwise upload the icon images.

I propose to address this as part of #467.

**Comment by tschaffter** [comment]
## Thomas' first review (WIP)

> The icons have not been implemented. Replace icon images by using the material icons, otherwise upload the icon images.

See Verena's [comment](https://github.com/Sage-Bionetworks/challenge-registry/issues/467#issuecomment-1201482317). Material icons are used successfully in the user account menu.

> How would I get the box-shadow or border styles (border px, border color) information from figma? I would need to fix missing border/box-shadow of challenge/organization card, i.e .challenge-card-container1, the box of challenge card

I will leave this question to @vpchung.

### Comments

- The default size of the user profile page looks like if I had zoom in. Here is what the page looks like by default:

image

image

- The buttons in the menu should be clickable anywhere in the "hover" zone. Currently the user has to click on the button text.

image

**Comment by vpchung** [comment]
> How would I get the box-shadow or border styles (border px, border color) information from figma? I would need to fix missing border/box-shadow of challenge/organization card, i.e .challenge-card-container1, the box of challenge card

Borders and shadows need to be implemented in teleportHQ, since the conversion isn't able to pick that up from Figma to teleportHQ.  

That being said, I thought I had done that before dumping the files - sorry about that!  I can add the properties in.

**Comment by tschaffter** [comment]
@vpchung @rrchai 

This branch has been updated with the latest content from `main`. After pulling the changes, you will need to rebuild your devcontainer because of a recent change made to it.

File: libs/challenge-registry/ui/src/lib/challenge-card/challenge-card.component.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    if (this.challenge) {
+      this.status = this.challenge.status ? this.challenge.status : 'No Status';
+      this.difficulty = this.challenge.difficulty
+        ? this.challenge.difficulty.replace(/([a-z])([A-Z])/g, '$1 $2')
+        : undefined;
+    }
+  }
+
+  printCamel(camel: string | undefined) {
+    return camel ? camel.replace(/([a-z])([A-Z])/g, '$1 $2') : 'No Status';
```
Comment (rrchai) [inline libs/challenge-registry/ui/src/lib/challenge-card/challenge-card.component.ts:26]: improvement: create a pipe will be used in html for future.

File: libs/challenge-registry/user-profile/src/lib/user-profile.component.html [hunk 2]
```diff
@@ -1,27 +1,60 @@
...
+        <span class="base-text03">ROCC user since</span>
+        <span class="base-text04">01/01/2020</span>
+      </div>
+      <div class="base-updatedat">
+        <span class="base-text05">Profile last updated</span>
+        <span class="base-text06">03/31/2022</span>
+      </div>
+      <challenge-registry-user-profile-stats
+        class="base-basic-stats"
+        [loggedIn]="loggedIn"
```
Comment (rrchai) [inline libs/challenge-registry/user-profile/src/lib/user-profile.component.html:24]: - if not loggin, send email the user.
- if loggin and it's different user, send email the user.
- if loggin and it's the same user, enable to edit

**Review by tschaffter — APPROVED** [review]

File: libs/challenge-registry/user-profile/src/lib/user-profile.component.html (hunk 2 above)
Comment (rrchai) [reply libs/challenge-registry/user-profile/src/lib/user-profile.component.html:24]: #538

**Pull request merged**